File: 21557 - Mukesh J/Week-5/547-number-of-provinces/547-number-of-provinces.java
class Solution {
    int[] parent;
    int[] rank;
    
    public int find(int m)
    {
        if(parent[m]==m)
            return m;
        return parent[m]=find(parent[m]);
    }
    
    public boolean union(int m, int n)
    {
        int parm = find(m);
        int parn = find(n);
        if(parm!=parn)
        {
            if(rank[parm]>rank[parn])
                parent[parn]=parm;
            else if(rank[parn]>rank[parm])
                parent[parm]=parn;
            else
           {
                parent[parn]=parm;
                rank[parm]++;
           }
            return true;
        }
        return false;
    }
    
    public int findCircleNum(int[][] isConnected) {
        parent= new int[isConnected.length+1];
        rank = new int[isConnected.length+1];
        int m,n;
        for(int j=0;j<parent.length;j++)
        {
            parent[j]=j;
            rank[j]=0;
        }
        
        int count= isConnected.length;
        for(int i=0;i<isConnected.length;i++)
        {
            for(int k=i+1;k<isConnected[i] .length;k++)
            {
                if(isConnected[i][k]==1)
                {
                    if(union(i,k))
                        count-=1;
                }
                
            }
        }
        return count;
    }
}

File: 21553 - Sai Nandan D/Week 5/547_Number of Provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        self.provinces = n
        
        parent = [i for i in range(n)]
        rank = [0] * n
        
        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]
        
        def union(a, b):
            x = find(a)
            y = find(b)
            
            if x != y:
                if rank[x] > rank[y]:
                    parent[y] = x
                elif rank[y] > rank[x]:
                    parent[x] = y
                else:
                    parent[y] = x
                    rank[x] += 1
                self.provinces -= 1
                
        for i in range(n):
            for j in range(i+1, n):
                if isConnected[i][j]:
                    union(i, j)
                
        return self.provinces
